src/main/java/org/spongepowered/common/data/builder/tileentity/AbstractTileEntityBuilder.java

```diff
@@ -0,0 +1,97 @@
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.block.BlockType;
+import org.spongepowered.api.block.entity.BlockEntity;
+import org.spongepowered.api.block.entity.BlockEntityType;
+import org.spongepowered.api.data.persistence.AbstractDataBuilder;
+import org.spongepowered.api.data.persistence.DataBuilder;
+import org.spongepowered.api.data.persistence.DataContainer;
+import org.spongepowered.api.data.persistence.DataView;
+import org.spongepowered.api.data.persistence.InvalidDataException;
+import org.spongepowered.api.data.persistence.Queries;
+import org.spongepowered.api.world.server.ServerWorld;
+import org.spongepowered.common.util.Constants;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public abstract class AbstractTileEntityBuilder<T extends BlockEntity> extends AbstractDataBuilder<T> implements DataBuilder<T> {
+    private final Supplier<BlockType> defaultBlockType;
+
+    protected AbstractTileEntityBuilder(Class<T> requiredClass, Supplier<BlockType> defaultBlockType, int supportedVersion) {
+        super(requiredClass, supportedVersion);
+        this.defaultBlockType = defaultBlockType;
+    }
+
+    @Override
+    protected Optional<T> buildContent(final DataView container) throws InvalidDataException {
+        Preconditions.checkNotNull(container, "DataContainer cannot be null in deserialization");
+
+        if (!this.validate(container)) {
+            return Optional.empty();
+        }
+
+        String worldName = container.getString(Constants.TileEntity.WORLD).orElseThrow(() -> new InvalidDataException("The provided container references a world that does not exist!"));
+        ServerWorld serverWorld = Sponge.getGame().getServer().getWorldManager().getWorld(worldName).orElseThrow(() -> new InvalidDataException("The provided container references a world that does not exist!"));
+
+        final BlockEntityType type = container.getCatalogType(Constants.TileEntity.TILE_TYPE, BlockEntityType.class).orElseThrow(() -> new InvalidDataException("The provided container references a block entity type that does not exist!"));
+        final BlockType blockType = this.defaultBlockType.get();
+
+        // Now we should be ready to actually translate the TileEntity with the right block.
+        final int x = container.getInt(Constants.DataSerializers.X_POS).get();
+        final int y = container.getInt(Constants.DataSerializers.Y_POS).get();
+        final int z = container.getInt(Constants.DataSerializers.Z_POS).get();
+
+        // If a plugin or mod cancels this, then return
+        if (!serverWorld.getLocation(x, y, z).setBlockType(blockType)) {
+            return Optional.empty();
+        }
+
+        final BlockPos pos = new BlockPos(x, y, z);
+        TileEntity tileEntity = ((net.minecraft.world.World) serverWorld).getTileEntity(pos);
+        if (tileEntity == null) {
+            return Optional.empty(); // TODO throw exception maybe?
+        }
+
```

Comment:
Also, this might serve as a suitable dynamic builder for all `TileEntity` types since `TileEntityType.create` allows creating the tile entity for us, and reading the tag compound (so we can stop relying on keeping our constants up to date and builders for the tile entities specifically, rather, we can generate them and register them as the `TileEntityType`s are registered).

Reply:
I could see how that would make sense. Of course this would probably build off the queries for the identifier and location as most of the nbt getters return an empty or default if something isn't present